src/pre_training.rs

```diff
@@ -0,0 +1,337 @@
+use itertools::Itertools;
+use ndarray::Array1;
+use std::collections::HashMap;
+use std::iter::Iterator;
+
+use crate::FSRSItem;
+
+pub fn pretrain(fsrs_items: Vec<FSRSItem>) -> [f32; 4] {
+    let pretrainset = create_pretrain_data(fsrs_items);
+    let rating_count = total_rating_count(pretrainset.clone());
+    let rating_stability = search_parameters(pretrainset);
+    smooth_and_fill(&mut rating_stability.clone(), &rating_count)
+}
+
+fn create_pretrain_data(fsrs_items: Vec<FSRSItem>) -> HashMap<i32, HashMap<String, Vec<f32>>> {
+    // filter FSRSItem instances with exactly 2 reviews.
+    let items: Vec<_> = fsrs_items
+        .into_iter()
+        .filter(|item| item.reviews.len() == 2)
+        .collect();
+
+    // use a nested HashMap (groups) to group items first by the rating in the first FSRSReview
+    // and then by the delta_t in the second FSRSReview.
+    let mut groups: HashMap<i32, HashMap<i32, Vec<i32>>> = HashMap::new();
+
+    for item in items {
+        let first_rating = item.reviews[0].rating;
+        let second_delta_t = item.reviews[1].delta_t;
+        let second_label = if item.reviews[1].rating == 1 { 0 } else { 1 };
+
+        let inner_map = groups.entry(first_rating).or_insert_with(HashMap::new);
+        let ratings = inner_map.entry(second_delta_t).or_insert_with(Vec::new);
+        ratings.push(second_label);
+    }
+
+    let mut results = HashMap::new();
+
+    for (first_rating, inner_map) in &groups {
+        let mut data = vec![]; // (delta_t, recall, count)
+
+        // calculate the average (recall) and count for each group.
+        for (second_delta_t, ratings) in inner_map {
+            let avg: f64 = ratings.iter().map(|&x| x as f64).sum::<f64>() / ratings.len() as f64;
+            data.push((*second_delta_t, avg, ratings.len()));
+        }
+
+        // Sort by delta_t in ascending order
+        data.sort_by(|a, b| a.0.cmp(&b.0));
+
+        let delta_t: Vec<_> = data.iter().map(|&(dt, _, _)| dt as f32).collect();
+        let recall: Vec<_> = data.iter().map(|&(_, r, _)| r as f32).collect();
+        let count: Vec<_> = data.iter().map(|&(_, _, c)| c as f32).collect();
+
+        let mut inner_result = HashMap::new();
+        inner_result.insert("delta_t".to_string(), delta_t);
+        inner_result.insert("recall".to_string(), recall);
+        inner_result.insert("count".to_string(), count);
```

Comment:
Could this be moved into a separate struct, instead of using untyped keys?

Otherwise, nice improvement :-)

Reply:
The rest looks good, but I still think this should use a Struct instead of a HashMap - is there any reason it cannot?